codex-rs/mcp-server/src/conversation_loop.rs

```diff
@@ -1,121 +1,444 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::path::PathBuf;
 use std::sync::Arc;
 
-use crate::exec_approval::handle_exec_approval_request;
-use crate::outgoing_message::OutgoingMessageSender;
-use crate::outgoing_message::OutgoingNotificationMeta;
-use crate::patch_approval::handle_patch_approval_request;
 use codex_core::Codex;
 use codex_core::protocol::AgentMessageEvent;
 use codex_core::protocol::ApplyPatchApprovalRequestEvent;
+use codex_core::protocol::Event;
 use codex_core::protocol::EventMsg;
 use codex_core::protocol::ExecApprovalRequestEvent;
+use codex_core::protocol::FileChange;
 use mcp_types::RequestId;
+use tokio::sync::Mutex;
+use tokio::sync::watch::Receiver as WatchReceiver;
 use tracing::error;
+use uuid::Uuid;
+
+use crate::exec_approval::handle_exec_approval_request;
+use crate::mcp_protocol::CodexEventNotificationParams;
+use crate::mcp_protocol::ConversationId;
+use crate::mcp_protocol::InitialStateNotificationParams;
+use crate::mcp_protocol::InitialStatePayload;
+use crate::mcp_protocol::NotificationMeta;
+use crate::outgoing_message::OutgoingMessageSender;
+use crate::patch_approval::handle_patch_approval_request;
 
+/// Deferred elicitation requests to be sent after InitialState when
+/// streaming is enabled. Preserves original event order (FIFO).
+enum PendingElicitation {
```

Comment:
Is there a way we can accomplish what we're trying to do here without having to duplicate so much logic down to the specific params of an Exec/Patch elicitation? 

I would imagine the InitialState would be managed by OutoingMessages and would be as close to the wire as possible so it would intercept things right before they go out so it can replay them with as little business logic as possible. Does that seem feasible here?